src/main/MainTabsScreen.js

```diff
@@ -119,3 +127,28 @@ export default function MainTabsScreen(props: Props) {
     </View>
   );
 }
+
+// TODO: Use `useSelector` instead of `connect`.
+//
+// The reason we can't do so right now is that, apparently, mixing
+// `connect` with `useSelector` in an app can lead to an ordering of
+// components' updates that isn't predictable just from the component
+// hierarchy. See discussion at
+//   https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/converting.20to.20Hooks/near/1111919.
+//
+// A crash-inducing problem can stem from this: we found that
+// `mapStateToProps` on a descendant of `MainTabsScreen` was running
+// -- and throwing an uncaught error -- on logout, without
+// `MainTabsScreen`'s early return on `!haveServerData` preventing
+// that from happening.
+//
+// Thankfully, we've found that placing a `connect` here will prevent
+// that from happening:
+//   https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/converting.20to.20Hooks/near/1112465.
+//
+// When our other `connect` callsites are gone, or at least the ones
+// in `MainTabsScreen`'s descendants, we expect that we can safely
+// switch this to a `useSelector`.
```

Comment:
I think the conclusion is actually that we'll likely want to keep this here even in a future when all its descendants are using `useSelector`, because it's doing something useful for us that `useSelector` doesn't do: it interposes a new `ReactReduxContext.Provider` component, which proxies subscriptions so that the descendant components only rerender if this one continues to say their subtree should be kept around.

I guess right now it seems to break as long as a relevant descendant uses `connect`, and works if it's converted to `useSelector` too. After investigation, I wound up puzzled that `useSelector` would work here when `connect` doesn't:
https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/converting.20to.20Hooks/near/1111970
So clearly there is something I still don't understand. But, the job `connect` is doing for us here is one that makes sense to me as being a useful one.

Reply:
Ah OK, thanks! If I haven't got it quite right in this revision (which borrows heavily from your comment here), I'd appreciate it if you'd jump in and make any necessary changes to set the record straight. 🙂